File: source/Modern/RC4.cs
using System.Text;
namespace Ciphers
{
    namespace Modern
    {

        public class RC4 : Cipher
        {
            private byte[] _S;
            private byte _i = 0;
            private byte _j = 0;

            public RC4(string key)
            {
                _key = key;
                _S = new byte[256];
                initialise();
            }

            private byte[] getKeyStreamBytes(int n)
            {
                byte[] result = new byte[n];
                byte index = 0;
                for (var k = 0; k < n; k++)
                {
                    _i = (byte)((_i + 1) % 256);
                    _j = (byte)((_j + _S[_i]) % 256);

                    var temp = _S[_i];
                    _S[_i] = _S[_j];
                    _S[_j] = temp;

                    index = (byte)((_S[_i] + _S[_j]) % 256);
                    result[k] = _S[index];
                }
                return result;
            }

            public override string Encode(string plain)
            {
                int len = plain.Length;
                byte[] msgBytes = new byte[len];
                byte[] keyBytes = getKeyStreamBytes(len);
                byte[] cipherBytes = new byte[len];
                for (int i = 0; i < len; i++)
                {
                    msgBytes[i] = ((byte)plain[i]);
                }

                for (int i = 0; i < len; i++)
                {
                    cipherBytes[i] = (byte)((keyBytes[i] ^ msgBytes[i]) % 256);
                }


                return BitConverter.ToString(cipherBytes);
            }

            public override string Decode(string cipher)
            {
                //reinitialise the PRNG
                initialise();
                List<byte> msgBytes = new List<byte>();
                StringBuilder res = new StringBuilder();
                foreach (var hex in cipher.Split("-"))
                {
                    msgBytes.Add(byte.Parse(hex, System.Globalization.NumberStyles.HexNumber));
                }
                byte[] keyBytes = getKeyStreamBytes(msgBytes.Count);
                for (int i = 0; i < msgBytes.Count; i++)
                {
                    res.Append((char)(msgBytes[i] ^ keyBytes[i]));
                }
                return res.ToString();
            }

            private void initialise()
            {
                _i = _j = 0;
                List<byte> byteKey = new List<byte>();

                //converting a string into a byte stream
                foreach (char symb in _key)
                {
                    byteKey.Add((byte)symb);
                }

                var len = byteKey.Count;
                //initialising the identity permutation
                for (int i = 0; i < 256; i++)
                {
                    _S[i] = (byte)i;
                }
                //initialising the permutation
                byte j = 0;
                for (int i = 0; i < 256; i++)
                {
                    j = (byte)(((j + _S[i] + _key[i % len])) % 256);

                    //swapping the values
                    //TODO: refactor into a method
                    var temp = _S[i];
                    _S[i] = _S[j];
                    _S[j] = temp;

                }
            }

        }
    }
}

File: src/ciphers/modern/RC4.go
package modern

import "encoding/hex"

type RC4 struct {
	key string
	s   []byte
	i   byte
	j   byte
}

func NewRC4(key string) *RC4 {
	S := make([]byte, 256)
	trc4 := RC4{
		key,
		S,
		0,
		0,
	}
	trc4.scheduleKey([]byte(key))
	return &trc4
}

func (R RC4) GetType() string {
	return "RC4, a symmetric stream cipher"
}

func (R RC4) Encode(plain string) string {
	msg := []byte(plain)
	msgLen := len(plain)
	keyBytes := R.getKeyStreamBytes(msgLen)
	res := make([]byte, msgLen)
	for i := 0; i < msgLen; i++ {
		res[i] = keyBytes[i] ^ msg[i]
	}
	return hex.EncodeToString(res)
}

func (R RC4) Decode(cipher string) string {
	R.scheduleKey([]byte(R.key))
	msg, ok := hex.DecodeString(cipher)
	if ok != nil {
		panic(ok)
	}
	msgLen := len(msg)
	keyBytes := R.getKeyStreamBytes(msgLen)
	res := make([]byte, msgLen)
	for i := 0; i < msgLen; i++ {
		res[i] = keyBytes[i] ^ msg[i]
	}
	return string(res)
}

func (R RC4) scheduleKey(key []byte) {
	keyLen := len(key)
	for i := 0; i < 256; i++ {
		R.s[i] = byte(i)
	}
	var j byte = 0
	for i := 0; i < 256; i++ {
		j = j + R.s[i] + key[i%keyLen]
		tmp := R.s[i]
		R.s[i] = R.s[j]
		R.s[j] = tmp
	}
}

func (R RC4) getKeyStreamBytes(n int) []byte {
	res := make([]byte, n)
	index := byte(0)
	for k := 0; k < n; k++ {
		R.i++
		R.j = R.j + R.s[R.i]

		tmp := R.s[R.i]
		R.s[R.i] = R.s[R.j]
		R.s[R.j] = tmp

		index = R.s[R.i] + R.s[R.j]
		res[k] = R.s[index]
	}
	return res
}
